OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,286 @@
+        return _dependentsMetaData;
+    }
+    
+    /** Set metadata associated with the dependent columns.                   */
+    void 
+    setDependentsMetaData(const DependentsMetaData& dependentsMetaData) {
+        _dependentsMetaData = dependentsMetaData;
+        validateDependentsMetaData();
+    }
+
+protected:
+    virtual size_t computeNumRows() const            = 0;
+    virtual size_t computeNumColumns() const         = 0;
+    virtual void validateIndependentMetaData() const = 0;
+    virtual void validateDependentsMetaData() const  = 0;
+
+    TableMetaData       _tableMetaData;
+    DependentsMetaData  _dependentsMetaData;
+    IndependentMetaData _independentMetaData;
+}; // AbstractDataTable
+
+
+/** DataTable_ is a in-memory storage container for data with support for 
+holding metadata. Data contains an independent column and a set of dependent 
+columns. The type of the independent column can be configured using ETX (
+template param). The type of the dependent columns, which together form a matrix
+, can be configured using ETY (template param). Independent and Dependent 
+columns can contain metadata. DataTable_ as a whole can contain metadata.     */
+template<typename ETX = double, typename ETY = SimTK::Real>
+class DataTable_ : public AbstractDataTable {
+public:
+    using RowVector    = SimTK::RowVector_<ETY>;
+    using ColumnVector = SimTK::Vector_<ETY>;
+
+    DataTable_()                             = default;
+    DataTable_(const DataTable_&)            = default;
+    DataTable_(DataTable_&&)                 = default;
+    DataTable_& operator=(const DataTable_&) = default;
+    DataTable_& operator=(DataTable_&&)      = default;
+    ~DataTable_()                            = default;
+
+    std::unique_ptr<AbstractDataTable> clone() const override {
+        return std::unique_ptr<AbstractDataTable>{new DataTable_{*this}};
+    }
+
+    /** Append row to the DataTable_.                                         */
+    void appendRow(const ETX& indRow, const RowVector& depRow) {
+        validateAppendRow(indRow, depRow);
+
+        _indData.push_back(indRow);
+
+        if(_depData.nrow() == 0 || _depData.ncol() == 0) {
+            try {
+                auto& labels = 
+                    _dependentsMetaData.getValueArrayForKey("labels");
+                if(depRow.size() != labels.size())
+                    throw Exception{"Number of columns in the input row does "
+                            "not match the number of labels."};
+            } catch(std::out_of_range&) {
+                // No operation.
+            }
+            _depData.resize(1, depRow.size());
+        } else
+            _depData.resizeKeep(_depData.nrow() + 1, _depData.ncol());
+            
+        _depData.updRow(_depData.nrow() - 1) = depRow;
+    }
+
+    /** Get row at index.                                                     */
+    RowVector getRowAtIndex(size_t index) const {
```

Comment:
This returns a copy of the row data -- is that what you want? If you change the return type to RowVectorView it will refer to the internal data instead (RowVectorView acts just like RowVector in most circumstances).